### PR TITLE
Replace asm memlcp() with C version

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

**Comment by beckerg** [comment]
See notes in ticket:  https://jira.micron.com/jira/browse/NFSE-5394

File: lib/cn/wbt_builder.h [hunk 1]
```diff
@@ -18,22 +18,21 @@ struct wbt_desc;
...
- * but can be faster if keys have long common prefixes.  It has no known issues
- * with alignment.
+ * on-media wastage of 3.5-bytes per key.  This approach consumes more storage
+ * and/or RAM, but can be faster if keys have long common prefixes.  It has no
+ * known alignment issues.
  */
-static_assert(HSE_KEY_CPE_ALIGNMENT == 4 || HSE_KEY_CPE_ALIGNMENT == 8,
-              "invalid alignment for structs key_state_entry_leaf and intern_key");
+static_assert(HSE_KEY_CPE_ALIGNMENT == 1 || HSE_KEY_CPE_ALIGNMENT == 8,
+              "invalid alignment for key common-prefix-elimination");
```
Comment (beckerg) [inline lib/cn/wbt_builder.h:35]: I'm going to remove all this, with the new memlcp() I can no longer produce any significant improvement to using memlcpq() here.

Comment (tristan957) [reply lib/cn/wbt_builder.h:35]: Will you remove it in this PR?

**Review by tristan957 — APPROVED** [review]

**Comment by beckerg** [comment]
Repushed to eliminate the key-cpe stuff.

**Comment by tristan957** [comment]
I will submit a follow-up PR to use GCC LTO for our CI LTO builds, unless people think we need both or only clang

File: lib/util/include/hse/util/arch.h [hunk 2]
```diff
@@ -162,8 +162,62 @@ get_time_ns(void)
...
+ * returning the maximum length at which they compare identical.
+ */
+static HSE_ALWAYS_INLINE size_t
+memlcp(const void *s1, const void *s2, size_t len)
+{
+    const uint8_t *lhs = s1;
+    const uint8_t *rhs = s2;
+    size_t i = 0;
+
+    while (i < (len & ~7ul)) {
```
Comment (nabeelmmd) [inline lib/util/include/hse/util/arch.h:181]: I'm sure there's a reason you chose not to precompute (len & ~7ul) prior to entering this while loop.. Curious to learn what that is...

Comment (beckerg) [reply lib/util/include/hse/util/arch.h:181]: Heh..  It should only be computed once prior to entering the loop, and I len mustn't change for the second loop.  Perhaps I should make len a const, that seems justifiable in this case :-)

**Review by nabeelmmd — APPROVED** [review]

**Pull request merged**